### PR TITLE
Unilm Model Support & GenerationMixinV3

### PR DESCRIPTION
**Pull request by fuliucansheng**
1. Unilm model implement refer to unilm repo & cache optimization for beam search.
2. GenerationMixinV3 (faster beam search for hf models) refer to fairseq repo, also support sampling methods.
3. Add unilm finetune weight trained by XSUM data for unittest.

### CODE REVIEW

**Comment by JiushengChen** [comment]
Please add it into our benchmark set so we can keep tracking its performance. Refer to "benchmarks/models/hf_bart.sh" for example.

File: fastseq/models/unilm_hf/__init__.py [hunk 1]
```diff
@@ -0,0 +1,6 @@
+from fastseq.models.unilm_hf.configuration_unilm import (
+    UnilmConfig,
+    UNILM_PRETRAINED_CONFIG_ARCHIVE_MAP,
+)
```
Comment (feihugis) [inline fastseq/models/unilm_hf/__init__.py:4]: Why import them here?

File: fastseq/models/unilm_hf/configuration_unilm.py [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+# The above copyright notice and this permission notice shall be included in all
+# copies or substantial portions of the Software.
+
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
```
Comment (feihugis) [inline fastseq/models/unilm_hf/configuration_unilm.py:22]: Could we use the same license header as other files in Fastseq like [here](https://github.com/microsoft/fastseq/blob/main/fastseq/models/__init__.py#L1-L2)?

File: fastseq/models/unilm_hf/configuration_unilm.py [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+""" UniLM model configuration """
+
+from __future__ import absolute_import, division, print_function, unicode_literals
+
+import json
+import logging
+import sys
```
Comment (feihugis) [inline fastseq/models/unilm_hf/configuration_unilm.py:29]: sort these imports in the alphabet order. Same to other places.

File: fastseq/models/unilm_hf/configuration_unilm.py [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+from io import open
+
+from transformers import PretrainedConfig
+
+logger = logging.getLogger(__name__)
```
Comment (feihugis) [inline fastseq/models/unilm_hf/configuration_unilm.py:34]: Let's create logger like [here](https://github.com/microsoft/fastseq/blob/main/fastseq/models/prophetnet_fs/bert_dictionary.py#L17-L19) so that we can configure logging in a single place. Same to other places.

File: fastseq/models/unilm_hf/modeling_unilm.py [hunk 5]
```diff
@@ -0,0 +1,770 @@
+from __future__ import absolute_import, division, print_function, unicode_literals
```
Comment (feihugis) [inline fastseq/models/unilm_hf/modeling_unilm.py:1]: add license header

**Comment by yuyan2do** [comment]
Consider add unilm finetune weight trained by CNNDM data. Easier to track speed across models.

File: fastseq_cli/transformers_generate.py [hunk 6]
```diff
@@ -5,9 +5,78 @@
...
+                    return tokenizer_class_fast.from_pretrained(pretrained_model_name_or_path, *inputs, **kwargs)
+                else:
+                    return tokenizer_class_py.from_pretrained(pretrained_model_name_or_path, *inputs, **kwargs)
+
+        raise ValueError(
+            "Unrecognized configuration class {} to build an AutoTokenizer.\n"
+            "Model type should be one of {}.".format(
+                config.__class__, ", ".join(c.__name__ for c in TOKENIZER_MAPPING.keys())
+            )
+        )
```
Comment (feihugis) [inline fastseq_cli/transformers_generate.py:78]: If users want to use `AutoModelForSeq2SeqLM` to load UniLM, they need to use this class instead of the one in huggingface, right? If yes, could we use `@replace` decorator to replace the ones in huggingface? So that no matter which one users import, they can always load UniLM. And these two classes can be moved to `fastseq/models/modeling_auto_hf.py`

File: tests/models/test_unilm_hf.py [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+                  encoding="utf-8") as expected_output_file:
+            for line in expected_output_file:
+                self.expected_outputs.append(line.strip())
+
+    @parameterized.named_parameters({
+        'testcase_name': 'Normal',
+        'beam_size': 5,
+        'batch_size': 128,
+        'lenpen': 1.0,
+        'no_repeat_ngram_size': 0
```
Comment (feihugis) [inline tests/models/test_unilm_hf.py:58]: why use no_repeat_ngram_size=0 here?

**Review by feihugis — COMMENTED** [review]
@fuliucansheng Thanks so much for your contribution! I did a pass and left some comments. I'm wondering if you can integrate the optimizations you made in GenerationMininV3 to [GenerationMininV2](https://github.com/microsoft/fastseq/blob/main/fastseq/optimizer/transformers/beam_search_optimizer.py#L64)?

**Comment by feihugis** [comment]
Maybe we can also add a README file for UniLM under `fastseq/model/unilm_hf` to introduce UniLM.

**Comment by JiushengChen** [comment]
> Maybe we can also add a README file for UniLM under `fastseq/model/unilm_hf` to introduce UniLM.

Let's create a folder  "examples/unilm", add README there. There are already a bunch of model examples there.

**Comment by JiushengChen** [comment]
CI test failed. @NickNickGo, please help take a look.

**Comment by NickNickGo** [comment]
> CI test failed. @NickNickGo, please help take a look.

I think this change doesn't pass transformer tests. Following is assertion error.

https://dev.azure.com/adsbrain/Fastseq/_build/results?buildId=1129&view=logs&j=18d0c406-0431-55f5-b805-33b8f79ed98b&t=4bb5bc85-2986-577f-db09-26b4ec255899&l=32058

**Comment by feihugis** [comment]
> > CI test failed. @NickNickGo, please help take a look.
> 
> I think this change doesn't pass transformer tests. Following is assertion error.
> 
> https://dev.azure.com/adsbrain/Fastseq/_build/results?buildId=1129&view=logs&j=18d0c406-0431-55f5-b805-33b8f79ed98b&t=4bb5bc85-2986-577f-db09-26b4ec255899&l=32058

@NickNickGo @JiushengChen  I have synced with @fuliucansheng offline. He is working on a fix.

**Comment by fuliucansheng** [comment]
I have resolved all the comments and passed all checks. Great thanks to @feihugis @JiushengChen @NickNickGo for help. ^ ^

File: fastseq/optimizer/transformers/beam_search_optimizer.py [hunk 8]
```diff
@@ -717,12 +719,17 @@ def _generate_beam_search(
...
+
+        cand_size = 2 * num_beams
+        cand_offsets = torch.arange(0, cand_size).type_as(input_ids)
 
         # cache compute states
         past = (encoder_outputs, None) if encoder_outputs is not None else None
 
         # done sentences
         done = [False for _ in range(batch_size)]
+        use_reorder_cache_v2 = hasattr(self, '_reorder_cache_v2')
```
Comment (feihugis) [inline fastseq/optimizer/transformers/beam_search_optimizer.py:732]: In my understanding, the changes in `_generate_beam_search` are to remove the finished sentences from the batch during the iteration. As the batch size will be changed, the behavior of `_reorder_cache` needs to be changed as well. That's why `_reorder_cache_v2` is added. These changes got 10% ~ 40% speedup as said in the email.

Is my understanding correct? If yes, could we add some comments here like this: currently only UniLM supports the feature that removes the finished hypotheses from the batch during the generation. Other models in huggingfaces do not support it yet, which requires a new version `_reorder_cache_v2` to support the dynamic changed batch size. Once all the models support this feature, we will merge `_reorder_cache_v2` and `_reorder_cache` into one.

**Review by feihugis — COMMENTED** [review]
@fuliucansheng Thanks for the revisions. Left some minor comments. Could you also update the README file to introduce how to use UniLM (e.g., introduction, model checkpoint, training, fine-tuning, inference)? An example can be found [here](https://github.com/microsoft/prophetnet).

File: fastseq/optimizer/transformers/beam_search_optimizer.py (hunk 8 above)
Comment (fuliucansheng) [reply fastseq/optimizer/transformers/beam_search_optimizer.py:732]: Yeah, you are right. ` _reorder_cache` be reserved for other people who use the hf beam search.

Comment (feihugis) [reply fastseq/optimizer/transformers/beam_search_optimizer.py:732]: Great, please add a comment to the code. It will help developers read the code.

File: setup.py [hunk 9]
```diff
@@ -54,6 +54,7 @@ def get_fastseq_version():
         "packaging",
         "torch>=1.4.0",
         "pytorch-transformers==1.0.0",
+        "transformers==3.0.2"
```
Comment (JiushengChen) [inline setup.py:57]: I think it is already handled in extras_require

Comment (fuliucansheng) [reply setup.py:57]: I added this because the CI test reported an error of `No module named 'transformers'`. 
https://github.com/microsoft/fastseq/pull/80/checks?sha=2f3ef72e81fefd4a0e37f3ca29b8c5e1ba7367da

Comment (feihugis) [reply setup.py:57]: @fuliucansheng  As FastSeq supports both FairSeq and HuggingFace-Transformers, these two packages are added as the [extras_require](https://github.com/microsoft/fastseq/blob/main/setup.py#L18-L21). 

To handle the above issue, could you please handle it like [here](https://github.com/microsoft/fastseq/blob/main/fastseq/models/prophetnet_fs/__init__.py#L12-L23), which throws a warning message instead of an error when `transformers` is not installed.

Comment (JiushengChen) [reply setup.py:57]: And this error is strange "No module named 'transformers'". It should be already installed in CI. Any idea?

File: examples/unilm/README.md [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+```bash
+$ fastseq-generate-for-transformer \
+      cnndm-unilm-base-cased \
+	  path/to/cnndm_rawtext_input \
+	  path/to/cnndm_summary_output \
+      --fp16 \
+	  --bs 128 \
+	  --max_length 608 \
+	  --max_gen_length 160 \
+	  --no_repeat_ngram_size 3
```
Comment (feihugis) [inline examples/unilm/README.md:33]: Let's make the indent consistent.

Comment (feihugis) [reply examples/unilm/README.md:33]: Do you have any command examples for training and fine-tuning?

File: fastseq_cli/transformers_generate.py [hunk 11]
```diff
@@ -314,9 +331,15 @@ def run_generate():
...
         padding=args.padding,
+<<<<<<< HEAD
+        max_length=args.max_length,
+        max_gen_length=args.max_gen_length
+        )
+=======
         max_tokenizer_length=args.max_tokenizer_length,
         max_gen_length=args.max_gen_length)
 
+>>>>>>> 0b7afdfa74c27b7dde7e8633dba21c20b72344fe
```
Comment (feihugis) [inline fastseq_cli/transformers_generate.py:342]: Could you please resolve the conflicts here?

File: setup.py (hunk 9 above)
Comment (feihugis) [reply setup.py:57]: @JiushengChen The cause is [here](https://github.com/microsoft/fastseq/blob/main/tests/run_fairseq_tests.py#L26-L32): the unit tests of FairSeq run in another virtual environment, which does not install transformers.

**Review by feihugis — COMMENTED** [review]
@fuliucansheng Thanks for the revisions! Please find some minor comments in below.

File: examples/unilm/README.md (hunk 10 above)
Comment (fuliucansheng) [reply examples/unilm/README.md]: Add some demo training code for reference~ Additional code of preprocessing and trainer are required for e2e training.

File: fastseq_cli/transformers_generate.py [hunk 12]
```diff
@@ -173,6 +173,9 @@ def generate_summaries_or_translations(
     if decoder_start_token_id is None:
         decoder_start_token_id = gen_kwargs.pop("decoder_start_token_id", None)
 
+    if hasattr(tokenizer, 'model_max_length'):
+        tokenizer.model_max_length = max_tokenizer_length
```
Comment (feihugis) [inline fastseq_cli/transformers_generate.py:177]:
```suggestion
    if hasattr(tokenizer, 'model_max_length') and max_tokenizer_length != None:
        tokenizer.model_max_length = max_tokenizer_length
```

File: setup.py (hunk 9 above)
Comment (JiushengChen) [reply setup.py]: @feihugis , what is the solution?

**Review by feihugis — APPROVED** [review]

File: setup.py (hunk 9 above)
Comment (feihugis) [reply setup.py]: The solution is to throw a warning [here](https://github.com/microsoft/fastseq/pull/80/files#diff-d722765538278d903ef4b391db1a989ac3ba213f7a364b57e1ffba3ed1fcb14aR12-R16).

**Comment by fuliucansheng** [comment]
Great thanks to @feihugis for detailed review and comments. :blush:

File: examples/unilm/README.md [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+- [Open Source](https://github.com/microsoft/unilm)
+
+## Speedup by using FastSeq
+
+- CNN daily mail validation data, NVIDIA-2080TI
+
+  |       BatchSize      |       32      |        64       |      128       |
+  |:--------------------:|:-------------:|:---------------:|:--------------:|
+  |      unilm           |  - samples/s  |  1.2 samples/s  |      OOM       |
+  |   above + fastseq    |  - samples/s  |  8.4 samples/s  |  8.4 samples/s |
```
Comment (JiushengChen) [inline examples/unilm/README.md:15]: any idea why bs32 result is empty? bs128 speed is same as bs64, looks strange.

File: examples/unilm/README.md [hunk 14]
```diff
@@ -0,0 +1,57 @@
+# Unilm
+
+- Paper
+	- [UnilmV1](https://arxiv.org/abs/1905.03197)
+	- [UnilmV2](https://arxiv.org/abs/2002.12804)
+- [Open Source](https://github.com/microsoft/unilm)
+
+## Speedup by using FastSeq
+
+- CNN daily mail validation data, NVIDIA-2080TI
```
Comment (JiushengChen) [inline examples/unilm/README.md:10]: hf_unilm.sh comment says it is v100, looks different from here, typo? We may need to use V100 for benchmark as auto test is based on this sku.

File: examples/unilm/README.md (hunk 13 above)
Comment (fuliucansheng) [reply examples/unilm/README.md:15]: This result is from my local machine with 2080TI, which only have 11G memory. Both of the GPU usage with bs=64 and 128 is almost 100%. The difference of speed is very a little. This results is comparison of unilm offical repo with import fastseq or not.

Comment (fuliucansheng) [reply examples/unilm/README.md:15]: In addition, unilm offical repo has its own implementation of beam search, which is a little different from the one in huggingface. Which one is better as a baseline for unilm ? Unilm official repo or optimized-unilm in fastseq with GenerationMixin ?

File: examples/unilm/README.md (hunk 14 above)
Comment (fuliucansheng) [reply examples/unilm/README.md:10]: ` benchmark/models/hf_unilm.sh` is the script for unilm speed test in fastseq benchmark, it can be compared with other models in the repo. It's not the result for unilm with fastseq or not, in my opinion.

File: examples/unilm/README.md (hunk 13 above)
Comment (JiushengChen) [reply examples/unilm/README.md:15]: Should be the one from unilm official repo?

File: examples/unilm/README.md (hunk 14 above)
Comment (JiushengChen) [reply examples/unilm/README.md:10]: To confirm,  the numbers in  "benchmark/models/hf_unilm.sh" are from V100, right? Then it should be fine.

Comment (fuliucansheng) [reply examples/unilm/README.md:10]: Yeah, that's right.

File: examples/unilm/README.md (hunk 13 above)
Comment (fuliucansheng) [reply examples/unilm/README.md:15]: I think so. This table could be updated after running it with V100~ This may need the help of @feihugis .

**Review by JiushengChen — APPROVED** [review]

File: examples/unilm/README.md (hunk 13 above)
Comment (JiushengChen) [reply examples/unilm/README.md:15]: it is ok. let's update it in separate pr.

File: examples/unilm/README.md (hunk 14 above)
Comment (feihugis) [reply examples/unilm/README.md:10]: @fuliucansheng, why not use the results from V100 here?

File: examples/unilm/README.md (hunk 13 above)
Comment (feihugis) [reply examples/unilm/README.md:15]: Synced with @fuliucansheng offline. The numbers in hf_unilm.sh are copied from the template file. I will run the benchmarks and update the benchmark numbers in another PR.

**Comment by feihugis** [comment]
@fuliucansheng  Thanks a lot for your contribution! I will update the benchmark numbers in another PR.
